skipchain/skipchain.go

```diff
@@ -381,6 +390,95 @@ func (s *Service) StoreSkipBlockInternal(psbd *StoreSkipBlock) (*StoreSkipBlockR
 	return reply, nil
 }
 
+// sendForwardLinkRequest sends a request to the given until until either the forward-link is
+// created or there's not enough online nodes to get a valid signature.
+func sendForwardLinkRequest(ro *onet.Roster, req *ForwardSignature, reply *ForwardSignatureReply) (err error) {
+	cl := NewClient()
+
+	// Try as many time as it can until the faulty threshold is reached
+	// meaning it's impossible to get a valid signature.
+	retries := protocol.DefaultFaultyThreshold(len(ro.List)) + 1
+
+	for i := 0; i < retries; i++ {
+		// No random permutation as we need a given threshold anyway.
+		err = cl.SendProtobuf(ro.List[i], req, reply)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+// OptimizeProof creates missing forward links to optimize the proof of the block
+// at the given index.
+func (s *Service) OptimizeProof(req *OptimizeProofRequest) (*OptimizeProofReply, error) {
+	pr, err := s.db.GetProofForID(req.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	target := pr[len(pr)-1]
+	index := 0
+	newProof := Proof{}
+
+	for _, sb := range pr[:len(pr)-1] {
+		if sb.Index < index {
+			// Skip blocks thanks to the new forward-link.
+			continue
+		}
+
+		diff := math.Log(float64(target.Index - sb.Index))
+		base := math.Log(float64(target.BaseHeight))
+		// Target the maximum height authorized for this block.
+		h := int(math.Min(diff/base, float64(sb.Height-1)))
+
+		if h > 0 && len(sb.ForwardLink) <= h {
+			// We need to round the e^x operation because of the floating-point
+			// precision but unlike the previous division, this will always
+			// produce the index minus ε < 10^-9.
+			index = sb.Index + int(math.Round(math.Exp(float64(h)*base)))
+			to := pr.Search(index)
+			if to == nil {
+				return nil, fmt.Errorf("chain is inconsistent: block at index %d not found", index)
+			}
+
+			req := &ForwardSignature{
+				TargetHeight: h,
+				Previous:     sb.Hash,
+				Newest:       to,
+			}
+			reply := &ForwardSignatureReply{}
+
+			log.Lvlf2("Request to create forward-link at index %d with height %d / %d", sb.Index, h, index)
+			// The signature must be asked to the roster of the block
+			err := sendForwardLinkRequest(sb.Roster, req, reply)
+
+			if err != nil {
+				log.Error("couldn't create a forward link:", err)
```

Comment:
```suggestion
				log.Error("could not create a missing forward link:", err)
```

(reasoning: the error messages in this function should be distinct from the ones in normal skipblock creation)